src/Workspaces/Core/Portable/Remote/RemoteHostAssetSerialization.cs

```diff
@@ -0,0 +1,89 @@
+{
+    internal static class RemoteHostAssetSerialization
+    {
+        public static async Task WriteDataAsync(ObjectWriter writer, IRemotableDataService remotableDataService, int scopeId, Checksum[] checksums, CancellationToken cancellationToken)
+        {
+            writer.WriteInt32(scopeId);
+
+            // special case
+            if (checksums.Length == 0)
+            {
+                writer.WriteInt32(0);
+                return;
+            }
+
+            if (checksums.Length == 1)
+            {
+                var checksum = checksums[0];
+                var remotableData = (await remotableDataService.GetRemotableDataAsync(scopeId, checksum, cancellationToken).ConfigureAwait(false)) ?? RemotableData.Null;
+                writer.WriteInt32(1);
+
+                await WriteRemotableData(writer, checksum, remotableData, cancellationToken).ConfigureAwait(false);
+                return;
+            }
+
+            var remotableDataMap = await remotableDataService.GetRemotableDataAsync(scopeId, checksums, cancellationToken).ConfigureAwait(false);
+            writer.WriteInt32(remotableDataMap.Count);
+
+            foreach (var (checksum, remotableData) in remotableDataMap)
+            {
+                await WriteRemotableData(writer, checksum, remotableData, cancellationToken).ConfigureAwait(false);
+            }
+
+            static async Task WriteRemotableData(ObjectWriter writer, Checksum checksum, RemotableData remotableData, CancellationToken cancellationToken)
+            {
+                checksum.WriteTo(writer);
+                writer.WriteInt32((int)remotableData.Kind);
+
+                await remotableData.WriteObjectToAsync(writer, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        public static Task<List<(Checksum, object)>> ReadDataAsync(Stream stream, int scopeId, ISet<Checksum> checksums, ISerializerService serializerService, CancellationToken cancellationToken)
+        {
+            var results = new List<(Checksum, object)>();
+
+            using var reader = ObjectReader.TryGetReader(stream, leaveOpen: true, cancellationToken);
+
+            // We only get a reader for data transmitted between live processes.
+            // This data should always be correct as we're never persisting the data between sessions.
+            Contract.ThrowIfNull(reader);
+
+            var responseScopeId = reader.ReadInt32();
+            Contract.ThrowIfFalse(scopeId == responseScopeId);
+
+            var count = reader.ReadInt32();
+            Contract.ThrowIfFalse(count == checksums.Count);
+
+            for (var i = 0; i < count; i++)
+            {
+                var responseChecksum = Checksum.ReadFrom(reader);
+                Contract.ThrowIfFalse(checksums.Contains(responseChecksum));
+
+                var kind = (WellKnownSynchronizationKind)reader.ReadInt32();
+
+                // in service hub, cancellation means simply closed stream
```

Comment:
not certain i get this.

Reply:
I don't either. This is not new code, just moved.